### PR TITLE
feat(package): support `less >= v3.0.0`

### LINKED ISSUE #239: Support for `less >= v3.0.0`
Less.js v3 was just released: https://github.com/less/less.js/blob/master/CHANGELOG.md

Using the webpack resolver won't work correctly:
```
 FAIL  test/index.test.js
  ✓ should compile simple less without errors (305ms)
  ✕ should resolve all imports (20ms)
  ✕ should add all resolved imports as dependencies (12ms)
  ✕ should resolve all imports from node_modules using webpack's resolver (12ms)
  ✕ should add all resolved imports as dependencies, including node_modules (13ms)
  ✓ should resolve aliases as configured (13ms)
  ✓ should add all resolved imports as dependencies, including aliased ones (16ms)
  ✓ should resolve all imports from the given paths using Less' resolver (10ms)
  ✓ should add all resolved imports as dependencies, including those from the Less resolver (9ms)
  ✓ should allow to disable webpack's resolver by passing an empty paths array (9ms)
  ✕ should not try to resolve import urls (11ms)
  ✓ should allow to import non-less files (20ms)
  ✕ should compile data-uri function (15ms)
  ✕ should transform urls (20ms)
  ✕ should generate source maps with sourcesContent by default (13ms)
  ✕ should be possible to override sourceMap.outputSourceFiles (14ms)
  ✓ should install plugins (11ms)
  ✓ should not alter the original options object (9ms)
  ✕ should fail if a file is tried to be loaded from include paths and with webpack's resolver simultaneously (10ms)
  ✕ should provide a useful error message if the import could not be found (15ms)
  ✓ should provide a useful error message if there was a syntax error (11ms)
  ✕ should add a file with an error as dependency so that the watcher is triggered when the error is fixed (10ms)
  ✓ should be able to import a file with an absolute path (11ms)
```

### LINKED ISSUE #241: Less Loader installation peer dependencies
I have installed `"less": "^3.0.1",` and also its in my `package.json` but each time i install `less-loader` i got the error that `less-loader@4.0.5 requires a peer of less@^2.3.1 but none is installed. You must install peer dependencies yourself.`

Now i dont undertstand what is the main issue here again. In my `package.json` i now have 

> "less": "^3.0.1",

> "less-loader": "^4.0.5",

### PR DESCRIPTION
**Pull request by thorn0**
### `Issues`

- Closes #239 
- Closes #241

### CODE REVIEW

**Comment by evilebottnawi** [comment]
@thorn0 look very good, thanks!

**Comment by thorn0** [comment]
For the full test coverage we'll have to run tests with both Less 2.x and 3.x. Because of this:
```diff
+        less.version[0] >= 3
+          ? options.ext && !tildePrefixedModuleName.test(filename)
+            ? this.tryAppendExtension(filename, options.ext)
+            : filename
+          : filename.replace(matchMalformedModuleFilename, '$1');
```

**Comment by evilebottnawi** [comment]
@thorn0 can you do this?

**Comment by thorn0** [comment]
Also the tests sometimes fail now because of the race condition in Less: https://github.com/less/less.js/pull/3170

**Comment by evilebottnawi** [comment]
@thorn0 i think we can fix this in other PR, but if solution is simple you can do this here :+1:

**Comment by thorn0** [comment]
Re running tests with Less 2. I can do it, but not sure when. Consider merging this as is and opening an issue.

**Comment by fraywing** [comment]
Let's submit this thing!

**Comment by davidpanzarella** [comment]
Any chance this will get submitted soon? @evilebottnawi  @jhnns @d3viant0ne @michael-ciniawsky

File: src/createWebpackLessPlugin.js [hunk 1]
```diff
@@ -14,6 +15,11 @@ const isLessCompatible = /\.(le|c)ss$/;
 // However, we must not match something like @import "~some-module/file.less";
 const matchMalformedModuleFilename = /(~[^/\\]+)\.less$/;
 
+// This somewhat changed in Less 3.x. Now the file name comes without the
+// automatically added extension whereas the extension is passed in as `options.ext`.
+// So, if the file name matches this regexp, we simply ignore the proposed extension.
+const tildePrefixedModuleName = /^~[^/\\]+$/;
```
Comment (michael-ciniawsky) [inline src/createWebpackLessPlugin.js]: `tildePrefixedModuleName` => `isModule || isModuleName || isModuleRequest`

File: src/createWebpackLessPlugin.js [hunk 2]
```diff
@@ -33,8 +39,22 @@ function createWebpackLessPlugin(loaderContext) {
...
+    // Sync resolving is used at least by the `data-uri` function.
+    // This file manager doesn't know how to do it, so let's delegate it
+    // to the default file manager of Less.
+    // We could probably use loaderContext.resolveSync, but it's deprecated,
+    // see https://webpack.js.org/api/loaders/#this-resolvesync
+    supportsSync() {
+      return false;
+    }
+
+    loadFile(filename, currentDirectory, options /* , environment */) {
```
Comment (michael-ciniawsky) [inline src/createWebpackLessPlugin.js]: Either `enviroment` is needed or not, but please don't leave commented stuff

**Comment by michael-ciniawsky** [comment]
Seems to be a semver patch

**Comment by thorn0** [comment]
It's backward compatible. Works with both Less 2.x and 3.x. I forgot to change the `peerDependencies` field in `package.json1 though.

**Comment by michael-ciniawsky** [comment]
```diff
- "less": "^2.3.1",
+ "less": "^2.0.0 || ^3.0.0",
```

**Comment by thorn0** [comment]
Why not `"less": "^2.3.1 || ^3.0.1"`? Are you sure it'll work with say 2.0.0?

**Comment by michael-ciniawsky** [comment]
Only the major version is important here imho, this avoids noisy false warnings within the specified semver major ranges, e.g would `=< v2.3.1` currently not work ? If a certain minor/path version doesn't work, then this would needed to be reflected in the `peerDependencies` of course, but assuming this isn't the case here :)

**Comment by thorn0** [comment]
I respectfully disagree. Minor version means added features. Are you sure we don't use any features added between 2.0 and 2.3? You're not, so why change this?

**Comment by michael-ciniawsky** [comment]
Sure you can keep the status quo and just add `|| ^3.0.0` aswell

**Comment by michael-ciniawsky** [comment]
@thorn0 Thx

**Comment by fraywing** [comment]
@thorn0, go for it!

**Comment by thorn0** [comment]
@michael-ciniawsky done, pls have a look

File: test/helpers/compile.js [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
 
 const fixturePath = path.resolve(__dirname, '..', 'fixtures');
 const outputPath = path.resolve(__dirname, '..', 'output');
 
 function compile(fixture, moduleRules, resolveAlias = {}) {
   return new Promise((resolve, reject) => {
     const entry = path.resolve(fixturePath, 'less', `${fixture}.less`);
-
-    webpack({
+    const [majorVersion] = webpackPackageJson.version.split('.');
```
Comment (michael-ciniawsky) [inline test/helpers/compile.js]: Please drop this, I appreciate that you want to fix it, but adding `webpack >= v4.0.0` to the CI setup will be part of #233 asap. The loader works with the `webpack >= v4.0.0` Loader API (no affecting breaking changes between `v3.0.0` => `v4.0.0`)

Comment (thorn0) [reply test/helpers/compile.js]: Are you sure? The tests don't pass without this change.

Comment (michael-ciniawsky) [reply test/helpers/compile.js]: Yep, I'm sure about that, please pin `webpack` to `^3.0.0` (https://github.com/webpack-contrib/less-loader/blob/master/package.json#L65)  instead

**Comment by hlehmann** [comment]
is it good to be merged ?

**Comment by davidpanzarella** [comment]
Any update on when this will be merged?

**Comment by michael-ciniawsky** [comment]
Released in [`v4.1.0`](https://github.com/webpack-contrib/less-loader/releases/tag/v4.1.0) 🎉

**Pull request merged**